docs/proposals/maintenance.md

```diff
@@ -0,0 +1,216 @@
+[here](http://releases.k8s.io/release-1.1/docs/proposals/maintenance.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Node Maintenance
+
+## Abstract
+
+A proposal for implementing safe node maintenance primitives, in stages.
+
+## Overview
+
+Nodes need periodic maintenance, most of which require partial or total
+unavailability to pods.  The cluster admin would like for this to happen (1)
+automatically, and (2) respecting any service quality constraints.
+
+Examples of maintenance:
+
+- unmount/remount filesystems to change mkfs params
+- remove from service but leave up for debugging
+- reboot for kernel or image upgrade
+- remove from service for hardware repair
+- gracefully reducing the size of the cluster
+- take down machines all machines in a rack for maintenance
+
+## Service Quality
+
+This may be thought of as the cluster admin offering a guarantee or SLO, or as the
+user specifying their requirements.  Around evictions, there are typically three
+parameters:
+
+- time between evictions (of pods in the same replica set)
+- minimum shard strength (minimum fraction of pods healthy)
+- eviction notice
+
+## Implementations
+
+In order of increasing complexity (and benefit).
+
+### Drain one node
+
+This is implemented in https://github.com/kubernetes/kubernetes/pull/16698
+
+### One-shot drain of multiple nodes
+
+In this case, we try to meet some specific service goals specified by the admin
+on the command line.  The specification of these parameters only when we run the
+tool lets us provide the possibility of safety while avoiding questions of where
+to configure and store such specification (treated further down).
+
+```
+kubectl drain \
+  --min-shard-strength=0.75 \
+  --min-seconds-between-evictions=900 \
+  --grace=900 $nodes
+```
+
+This would still be pretty simple.  We offer no guarantees about safety in the
+face of crashes or restarts, but we do meet the needs of admins who want
+something more sophisticated than `sleep`.  If kubectl was interrupted and then
+restarted with the same arguments, the end state would be exactly the same and
+we'd never break the shard strength rule, but the evictions might be more
+closely spaced.
+
+
+### Implementation of the above as a controller
+
+The parameters of a specific lame, drain, or unlame request could be pushed to
+the API server as a drain object.  The drain controller would have to manage
+multiple drains in flight at once.  If the drains have differing parameters,
+then more aggressive drains would generally starve safer ones.  While this could
+be a useful feature in emergencies, it will be a lot easier to reason about
+multiple drains in progress if they all share the same parameters, which would
+be configured via kube-env and would represent the cluster admin's SLO.
+
+The drain object would thus look something like this
```

Comment:
Why do you say "The modeling of this is the first example of an action being encoded as a rest resource"? If I create a ReplicationController, say with replicas=1, isn't that an action "create one pod" encoded as a rest resource?

The approach you suggested seems a little less declarative than what's suggested here. IIUC the approach you suggested would still need a rest resource to store the drain parameters; but instead of specifying a nodelist in that same resource, you'd request drains individually on each node you wanted to drain. Why do you think that is better? One problem is that I don't think specifying the nodes to drain via label selector would work in that approach. Also you'd have a lot of object writes (one to the subresource or field on each node).